Update falcon api

eanylin opened the pull request:
Expand on existing Shipyard Falcon API Endpoints

eanylin committed:
Merge pull request #2 from eanylin/helm_shipyard

Merge with Master

eanylin committed:
Expand the Falcon API Framework

Leverage the functionalities of airflow CLI and the
airflow-rest-api-plugin

eanylin committed:
Update Error Handling for Dag Trigger

The if/else condition should be based on the
error response code in this case

eanylin committed:
Add Dag Trigger Poll

Allows polling of dag status

mark-burnett commented on shipyard_airflow/control/airflow_connections.py:
```diff
@@ -0,0 +1,122 @@
+
+    def on_get(self, req, resp, action, conn_id, protocol, host, port):
+        # Retrieve URL
+        web_server_url = self.retrieve_config('BASE', 'WEB_SERVER')
```
This is not an issue for this PR, but `retrieve_config` should probably not just return a string when it fails; instead, I think it should raise an exception.

eanylin replied:
Agree.  Have change codes to raise exception instead.

mark-burnett commented on shipyard_airflow/control/airflow_connections.py:
```diff
@@ -0,0 +1,122 @@
+        web_server_url = self.retrieve_config('BASE', 'WEB_SERVER')
+
+        if 'Error' in web_server_url:
+            resp.status = falcon.HTTP_400
```
In the case that the config file is missing (which is how this is triggered, I think), the error code should be `500`, because it is a server error, not a client error.

eanylin replied:
Agree that this should be HTTP 500 instead of 400 Error Code.  Have updated the codes.

mark-burnett commented on shipyard_airflow/control/airflow_connections.py:
```diff
@@ -0,0 +1,122 @@
+        else:
+            if action == 'add':
+                # Concatenate to form the connection URL, Remove whitespaces
+                conn_uri = "%s %s %s %s %s" % (protocol, '://', host, ':', port)
```
You might find [`urllib.parse`](https://docs.python.org/3/library/urllib.parse.html) to be pretty useful here.

eanylin replied:
Agree.  Have made the required changes to use urllib instead.

mark-burnett commented on shipyard_airflow/control/airflow_connections.py:
```diff
@@ -0,0 +1,122 @@
+       
+            # Return output
+            if response["output"]["stderr"]:
+                resp.status = falcon.HTTP_400
```
It's not clear to me what this error code should be.  Would retrying the request ever be expected to succeed?

eanylin replied:
The idea that I have is that the codes is wrapping around CLI which is usually triggered once and returns error if it fails (be it syntax or actual failure that is reflected in stderr).  Hence I was not really looking at retry in this case.  That is the reason behind using HTTP 400.

mark-burnett replied:
We probably want to keep retrying in mind everywhere going forward, but I can see why we might not differentiate these kinds of failures for now.

eanylin replied:
Agree.  I am actually thinking of using the retry functionalities in Airflow Dag for retry scenarios.

mark-burnett reviewed (changes requested):
I've suggested a few things that could help reduce the boilerplate per resource here; however, after those changes you may find that a helper function or two will reduce the duplication further.

eanylin committed:
Update Falcon API

1) Change the content of shipyard.conf from Upper to Lower Case
2) Return Exception instead of string if shipyard.conf is missing
3) Return HTTP 500 instead of 400 if shipyard.conf is missing
4) Make use of urllib instead
